src/librustc/traits/error_reporting.rs

```diff
@@ -2564,6 +2648,11 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             }
             ObligationCauseCode::ImplDerivedObligation(ref data) => {
                 let parent_trait_ref = self.resolve_vars_if_possible(&data.parent_trait_ref);
+
+                if handle_async_await(*parent_trait_ref.skip_binder(), &data.parent_code) {
+                    return;
+                }
```

Comment:
Could you elaborate on which cases are supported by adding `handle_async_await` here? I can't work out from reading the diff which cases this would catch but the existing call in `note_obligation_cause` wouldn't?